server/src-lib/Hasura/GraphQL/Schema/Mutation.hs

```diff
@@ -0,0 +1,526 @@
+  :: forall m n r. (MonadSchema n m, MonadTableInfo r m, MonadRole r m)
+  => QualifiedTable
+  -> Maybe SelPermInfo
+  -> UpdPermInfo
+  -> m (Maybe (Parser 'Input n (RQL.ConflictClauseP1 UnpreparedValue)))
+conflictObject table selectPerms updatePerms = runMaybeT $ do
+  tableName        <- lift $ qualifiedObjectToName table
+  columnsEnum      <- MaybeT $ tableUpdateColumnsEnum table updatePerms
+  constraints      <- MaybeT $ tciUniqueOrPrimaryKeyConstraints . _tiCoreInfo <$> askTableInfo table
+  constraintParser <- lift $ conflictConstraint constraints table
+  whereExpParser   <- lift $ boolExp table selectPerms
+  let objectName = tableName <> $$(G.litName "_on_conflict")
+      objectDesc = G.Description $ "on conflict condition type for table " <>> table
+      constraintName = $$(G.litName "constraint")
+      columnsName    = $$(G.litName "update_columns")
+      whereExpName   = $$(G.litName "where")
+      fieldsParser = do
+        constraint <- RQL.CTConstraint <$> P.field constraintName Nothing constraintParser
+        columns    <- P.field columnsName Nothing $ P.list columnsEnum
+        whereExp   <- P.fieldOptional whereExpName Nothing whereExpParser
+        pure $ case columns of
+          [] -> RQL.CP1DoNothing $ Just constraint
+          _  -> RQL.CP1Update constraint columns preSetColumns $
+            BoolAnd $ catMaybes [whereExp, Just $ fmapAnnBoolExp partialSQLExpToUnpreparedValue $ upiFilter updatePerms]
+  pure $ P.object objectName (Just objectDesc) fieldsParser
+  where preSetColumns = partialSQLExpToUnpreparedValue <$> upiSet updatePerms
+
+conflictConstraint
+  :: forall m n r. (MonadSchema n m, MonadTableInfo r m)
+  => NonEmpty Constraint
+  -> QualifiedTable
+  -> m (Parser 'Both n ConstraintName)
+conflictConstraint constraints table = memoizeOn 'conflictConstraint table $ do
+  tableName <- qualifiedObjectToName table
+  constraintEnumValues <- for constraints \constraint -> do
+    name <- textToName $ getConstraintTxt $ _cName constraint
+    pure ( P.mkDefinition name (Just "unique or primary key constraint") P.EnumValueInfo
+         , _cName constraint
+         )
+  let enumName  = tableName <> $$(G.litName "_constraint")
+      enumDesc  = G.Description $ "unique or primary key constraints on table " <>> table
+  pure $ P.enum enumName (Just enumDesc) constraintEnumValues
+
+
+
+-- update
+
+-- | Construct a root field, normally called update_tablename, that can be used
+-- to update rows in a DB table specified by filters
+updateTable
+  :: forall m n r. (MonadSchema n m, MonadTableInfo r m, MonadRole r m, Has QueryContext r)
+  => QualifiedTable       -- ^ qualified name of the table
+  -> G.Name               -- ^ field display name
+  -> Maybe G.Description  -- ^ field description, if any
+  -> UpdPermInfo          -- ^ update permissions of the table
+  -> Maybe SelPermInfo    -- ^ select permissions of the table (if any)
+  -> m (Maybe (FieldParser n (RQL.AnnUpdG UnpreparedValue)))
```

Comment:
I think it'll be a good idea to explain why `Maybe FieldParser` is the return type of the `updateTable` function , instead of just `FieldParser`. If this is explained elsewhere in the code, then please ignore this suggestion.

Reply:
Or rather: if this is explained elsewhere in the code, let's make that explanation into a `Note` and refer to it here :-)

Reply:
Done, will push in a sec.

Reply:
Pushed; please resolve this if you think the comment is clear enough, ping this thread otherwise. :)